File: c_programming/array/n42_degree_of_an_array_697.c
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include "utils.h"

static int32_t degree_array(const int64_t *array, size_t len, size_t *out)
{
    int32_t ret = 0;
    int64_t *buffer = NULL;
    size_t *index = NULL;
    size_t i = 0, j = 0;
    size_t ind = 0;
    size_t min_index = 0;
    int64_t e = 0;
    size_t count = 0;
    size_t max_count = 0;

    UTILS_CHECK_LEN(len);
    UTILS_CHECK_PTR(array);
    UTILS_CHECK_PTR(out);

    buffer = (int64_t *)calloc(sizeof(int64_t), len);
    UTILS_CHECK_PTR(buffer);
    memcpy(buffer, array, sizeof(int64_t) * len);

    index = (size_t *)calloc(sizeof(size_t), len);
    UTILS_CHECK_PTR(index);

    for (i = 0; i < len; i ++) {
        index[i] = i;
    }

    for (i = 0; i < len - 1; i ++) {
        for (j = 0; j < len - i - 1; j ++) {
            if (buffer[j] > buffer[j + 1]) {
                utils_swap_int64(buffer + j, buffer + j + 1);
                utils_swap_size_t(index + j, index + j + 1);
            }
        }
    }

    utils_print_int64_array(buffer, len, "the buffer:");
    utils_print_size_t_array(index, len, "index is  :");

    for (i = 0, j = 0; i < len - 1; i ++) {
        if (buffer[i] == buffer[i + 1]) {
            count ++;
            if (count >= 1 && count > max_count) {
                ind = utils_int64_abs(index[i + 1] - index[j]);
                min_index = ind > min_index ? ind : min_index;
                LOG("the min_index is %zu, ind = %zu\n", min_index, ind);
            } else if (count >= 1 && count == max_count) {
                ind = utils_int64_abs(index[i + 1] - index[j]);
                min_index = ind < min_index ? ind : min_index;
                LOG("the min_index is %zu, ind = %zu\n", min_index, ind);
            }
            max_count = count > max_count ? count : max_count;
        } else {
            count = 0;
            j = i + 1;
        }
    }

    *out = min_index + 1;

finish:
    UTILS_SAFE_FREE(buffer);
    UTILS_SAFE_FREE(index);
    return ret;
}

int32_t main(void)
{
    int32_t ret = 0;
    int64_t array[] = {1,2,2,3,1,4,2};
    size_t len = 0;

    ret = degree_array(array, ARRAY_SIZE(array), &len);
    UTILS_CHECK_RET(ret);

    LOG("the result is %zu\n", len);

finish:
    return ret;
}

File: rust_programming/array/src/n42_degree_of_an_array_697.rs

// 1,2,2,3,1

use utils::num::min;

fn degree_array(vec: &[i64]) -> usize
{
    let mut count:usize = 0;
    let mut max_count:usize = 0;
    let mut ind:usize = 0;
    let mut min_index:usize = 0;
    let mut buffer:Vec<i64> = Vec::from(vec);
    let mut index:Vec<i64> = Vec::from(vec);
    let len:usize = vec.len();
    let mut i:usize = 0;
    let mut j:usize = 0;

    println!("the len is {len}");

    while i < len {
        index[i] = i as i64;
        i += 1;
    }

    for i in 0..len {
        for j in 0..len - 1 - i {
            if buffer[j] > buffer[j + 1] {
                println!("swap the index {j}, {}", j + 1);
                buffer.swap(j, j + 1);
                index.swap(j, j + 1);
            }
        }
    }

    println!("buffer is {:?}", buffer);
    println!("index  is {:?}", index);

    j = 0;
    for i in 0..len - 1 {
        if buffer[i] == buffer[i + 1] {
            count += 1;
            if (count >= 1) && (count > max_count) {
                ind = i64::abs((index[i + 1] as i64) - (index[j] as i64)) as usize;
                if (ind > min_index) {
                    min_index = ind;
                }
            } else if (count >= 1) && (count == max_count) {
                ind = i64::abs((index[i + 1] as i64) - (index[j] as i64)) as usize;
                if (ind < min_index) {
                    min_index = ind;
                }
            }
            if count > max_count {
                max_count = count;
            }
        } else {
            count = 0;
            j = i + 1;
        }
    }

    return min_index + 1;
}


pub fn test_degree_array()
{
    let vec1:Vec<i64> = vec![1,2,2,3,1];
    let mut n:usize = degree_array(&vec1);
    println!("the n1 is {n}");

    let vec2:Vec<i64> = vec![1,2,2,3,1,4,2];
    n = degree_array(&vec2);
    println!("the n2 is {n}");

}